source/AI.cpp

```diff
@@ -1945,7 +1945,10 @@ bool AI::MoveTo(Ship &ship, Command &command, const Point &targetPosition,
 	bool isFacing = (dp.Unit().Dot(angle.Unit()) > .95);
 	if(!isClose || (!isFacing && !shouldReverse))
 		command.SetTurn(TurnToward(ship, dp, 0.9999));
-	if(isFacing)
+	// Work with a slightly lower maximum velocity to avoid border cases.
+	double maxVelocity = ship.MaxVelocity() * ship.MaxVelocity() * .99;
+	if(isFacing && (velocity.LengthSquared() <= maxVelocity
+			|| dp.Unit().Dot(velocity.Unit()) < .95))
 		command |= Command::FORWARD;
```

Comment:
I believe that by stopping the forward command when close to max velocity, we're preventing afterburners from being used. Afterburners push you beyond your maximum velocity, but the AI will only use afterburners if the forward command is also sent.
https://github.com/endless-sky/endless-sky/blob/1d7b9203e683047b9ca63be997a7be3b7176ed2e/source/AI.cpp#L2357-L2361

Reply:
How would you like to account for this?

Reply:
Does forward thrust have any effect once max speed is reached if combined with AB?  i.e. will afterburners be less effective without forward specifically beyond max velocity?

I know forward thrust contributes to accelerating AB when below max velocity.

Reply:
> Does forward thrust have any effect once max speed is reached if combined with AB? i.e. will afterburners be less effective without forward specifically beyond max velocity?

It would, as the thrust applied is from thrusters + afterburners.

Thrusters applying acceleration:
https://github.com/endless-sky/endless-sky/blob/268a17a2a5d1afb1bdc8bbddbf7cee78c3c3fb80/source/Ship.cpp#L4295
Afterburners applying acceleration:
https://github.com/endless-sky/endless-sky/blob/268a17a2a5d1afb1bdc8bbddbf7cee78c3c3fb80/source/Ship.cpp#L4339

Reply:
Perhaps ShouldUseAfterburner should be included in the calculation but for player ships AB is never used unless player asks as far as I know.  For example, player ship will not AB while navigating to land on a planet.

Reply:
> How would you like to account for this?

Depending on the order of operations, we might be able to check if we have the afterburner command, and if so ignore the speed cap on the forward command. I imagine that the afterburner command is only evaluated after the forward command, though. Otherwise, we might be able to take some of the code that determines if the afterburner should fire and do those same checks here, so if the afterburner should fire then we'd still push the forward command.

Reply:
A naïve approach would be to check if ShouldUseAfterburner is true and ignore or raise the speed cap if it is.

Reply:
Raising speed cap feels more ideal so that AB shuts of at true max velocity

Reply:
In that case, perhaps we do something like this:
```
	double maxVelocity = ship.MaxVelocity(ShouldUseAfterburner(ship));
	maxVelocity *= maxVelocity * 0.99;
	if(isFacing && (velocity.LengthSquared() <= maxVelocity
			|| dp.Unit().Dot(velocity.Unit()) < .95))
		command |= Command::FORWARD;
```
Then in Ship, we change MaxVelocity to take a boolean parameter to check if it should include the afterburner thrust.
```
double Ship::MaxVelocity(bool withAfterburner) const
{
	// v * drag / mass == thrust / mass
	// v * drag == thrust
	// v = thrust / drag
	double thrust = attributes.Get("thrust");
	double afterburnerThrust = attributes.Get("afterburner thrust");
	return (thrust ? thrust + afterburnerThrust * withAfterburner : afterburnerThrust) / Drag();
}
```

Reply:
I’ll try this thanks.